Adds keyscan option to known_hosts module

jgrowl opened the pull request:
##### ISSUE TYPE

<!--- Pick one below and delete the rest: -->
- Feature Pull Request
##### COMPONENT NAME

known_hosts
##### ANSIBLE VERSION

<!--- Paste verbatim output from “ansible --version” between quotes below -->

```
ansible 2.1.0.0
  config file = /etc/ansible/ansible.cfg
  configured module search path = ['/usr/share/ansible']

```
##### SUMMARY

<!--- Describe the change, including rationale and design decisions -->

Sometimes it is useful to automatically perform a keyscan and add the result to known_hosts. I will usually be trusting the first time I try to connect to a new machine and want to accept it as a known host, but still want protection from MitM attacks from there on out. 

```
- known_hosts:
    name: '{{ ip_address }}'
    path: "{{ home }}/.ssh/known_hosts"
    keyscan: once
    keyscan_type: ecdsa
```

I added a **keyscan** module parameter that has the options _never_, _once_, and _always_
With the _once_ option, there will be a keyscan performed if an entry does not exist for that host yet. No change will be made if there already is an entry so that there will be a failure at runtime if the fingerprint on the server changes.

The _always_ option was added also to address PR #377 by @dokipen and adds comment about it being dangerous in production environments to address concerns by @mcv21. Maybe I should add a warning message that is displayed at runtime anytime 'always' option is used also?

_never_ option will always be the default so that a key should have to be provided by the user.

A **keyscan_type** module parameter was also added so that I can choose the type. This is defaulted to _rsa_

The code change is a little ugly, but I tried to leave the existing structure untouched as much as possible.

gregdek commented:
Thanks @jgrowl for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

cheshire137 commented on system/known_hosts.py:
```diff
@@ -51,6 +51,21 @@
+    - always option is vulnerable to MitM attacks and therefore should never be used in production environments.
+    choices: ["never", "once", "always"]
+    required: no
+    deault: never
```
Should this be `default`?

jgrowl commented:
Thanks @cheshire137, I fixed that typo.

jgrowl commented:
ready_for_review

gregdek commented:
Thanks @jgrowl. To the current maintainers, @mcv21 please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

mcv21 commented on system/known_hosts.py:
```diff
@@ -192,7 +220,7 @@ def sanity_check(module,host,key,sshkeygen):
-def search_for_host_key(module,host,key,path,sshkeygen):
+def search_for_host_key(module,host,key,path,sshkeygen,keyscan,keyscan_type,sshkeyscan):
     '''search_for_host_key(module,host,key,path,sshkeygen) -> (found,replace_or_add,found_line)
 
```
The documentation should match the function prototype

mcv21 commented on system/known_hosts.py:
```diff
@@ -210,16 +238,20 @@ def search_for_host_key(module,host,key,path,sshkeygen):
                                  check_rc=False)
     if stdout=='' and stderr=='' and (rc==0 or rc==1):
-        return False, False, None #host not found, no other errors
+        keyscan_key = _keyscan(module, host, keyscan_type, sshkeyscan)
```
This will call ssh-keyscan even if keyscan is never, won't it?

mcv21 commented:
I've made a couple of comments on the code itself.

I think we should not support the ssh-keyscan = always option. In what circumstances is it anything other than a terrible idea?

mcv21 commented:
needs_revision

gregdek commented:
Thanks @jgrowl for this PR. This PR requires revisions, either because it fails to build or by reviewer request. Please make the suggested revisions. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

[This message brought to you by your friendly Ansibull-bot.]

jgrowl commented:
@mcv21 Thanks for the feedback. I only added in the 'always' option since it seemed to go along with 'never' and 'once' options. I was not planning on using it and it would only be for quick and dirty scripts. I suppose anyone truly wanting to always accept the public key can just choose to ignore known hosts checking with ssh args.

I removed the always option which lets me simplify the code that has very little impact on the existing structure.

I reverted the search_for_host_key method back so now the prototype matches. ssh-keyscan only gets called in the case where the user wants to keyscan and no entry is found immediately following search_for_host_key.

jgrowl commented:
ready_for_review

gregdek commented:
Thanks @jgrowl. To the current maintainers, @mcv21 please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

mcv21 commented:
shipit

Thanks for your work on this @jgrowl, and your patience!

gregdek commented:
Thanks again to @jgrowl for this PR, and thanks @mcv21 for reviewing. Marking for inclusion.

[This message brought to you by your friendly Ansibull-bot.]

alikins commented on system/known_hosts.py:
```diff
@@ -91,19 +105,31 @@ def enforce_state(module, params):
+
+        rc, out, err = module.run_command([sshkeyscan,'-t',keyscan_type,host], check_rc=True)
+        if out=='': # host not found
+            module.fail_json(msg="Unable to ssh-keyscan host {}".format(host))
```
.format() is not pythom2.4 compatible.

```
[fedora-23:ansible-modules-extras (pr-2488 %)]$ pyqver2 -m 2.5 -l system/known_hosts.py
system/known_hosts.py:130: 2.6 string literal .format()
```

alikins commented:
The .format() use needs to be removed to be py2.4/py2.5 compatible.

gregdek commented:
@jgrowl A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]

jgrowl commented:
ready_for_review

I'm getting a Shippable failure but I think it may be a build configuration issue and not related to my changes. I'm not sure though as I'm not that familiar with shippable!

gregdek commented:
Thanks @jgrowl. To the current maintainers, @mcv21 please review according to guidelines (http://docs.ansible.com/ansible/developing_modules.html#module-checklist) and comment with text 'shipit', 'needs_revision' or 'close_me' as appropriate.

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
Thanks @jgrowl for this PR. Unfortunately, it is not mergeable in its current state due to merge conflicts. Please rebase your PR. When you are done, please comment with text 'ready_for_review' and we will put this PR back into review.

For help on how to do this cleanly please see http://docs.ansible.com/ansible/community.html#contributing-code-features-or-bugfixes

[This message brought to you by your friendly Ansibull-bot.]

gregdek commented:
@jgrowl A friendly reminder: this pull request has been marked as needing your action. If you still believe that this PR applies, and you intend to address the issues with this PR, just let us know in the PR itself and we will keep it open pending your changes. When you do address the issues, please respond with ready_for_review in your comment, so that we can notify the maintainer.

[This message brought to you by your friendly Ansibull-bot.]
